src/transformStateWithClones.js

```diff
@@ -6,8 +6,50 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newObj = { ...state };
+  const stateHistory = [];
+
+  for (const { type, keysToRemove, extraData } of actions) {
+    switch (type) {
+      case 'addProperties':
+        addProperties(newObj, extraData);
+        stateHistory.push(Object.assign({}, newObj));
+        break;
+
+      case 'removeProperties':
+        removeProperties(newObj, keysToRemove);
+        stateHistory.push(Object.assign({}, newObj));
+        break;
+
+      case 'clear':
+        clearProperties(newObj);
+        stateHistory.push(Object.assign({}, newObj));
+        break;
+
+      default:
+        return 'some error';
```

Comment:
If action's type is not one of known ones (`addProperties`, `removeProperties`, `clear`), it makes sense to return the current state itself - so that the next actions would work correctly.